// @eclipse-glsp/client/lib/features/accessibility/toast/toast-handler.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShowToastMessageAction = exports.HideToastAction = void 0;
const sprotty_1 = require("@eclipse-glsp/sprotty");
var HideToastAction;
(function (HideToastAction) {
    HideToastAction.KIND = 'hideToastMessageAction';
    function is(object) {
        return sprotty_1.Action.hasKind(object, HideToastAction.KIND);
    }
    HideToastAction.is = is;
    function create(options) {
        return { kind: HideToastAction.KIND, options };
    }
    HideToastAction.create = create;
})(HideToastAction || (exports.HideToastAction = HideToastAction = {}));
var ShowToastMessageAction;
(function (ShowToastMessageAction) {
    ShowToastMessageAction.KIND = 'showToastMessageAction';
    ShowToastMessageAction.TIMEOUT = 2000;
    function is(object) {
        return sprotty_1.Action.hasKind(object, ShowToastMessageAction.KIND);
    }
    ShowToastMessageAction.is = is;
    function create(options) {
        var _a, _b;
        return { kind: ShowToastMessageAction.KIND, options: { ...options, position: (_a = options.position) !== null && _a !== void 0 ? _a : 'center', id: (_b = options.id) !== null && _b !== void 0 ? _b : Symbol('toast id') } };
    }
    ShowToastMessageAction.create = create;
    function createWithTimeout(options) {
        var _a, _b;
        return {
            kind: ShowToastMessageAction.KIND,
            options: { timeout: ShowToastMessageAction.TIMEOUT, ...options, position: (_a = options.position) !== null && _a !== void 0 ? _a : 'center', id: (_b = options.id) !== null && _b !== void 0 ? _b : Symbol('toast id') }
        };
    }
    ShowToastMessageAction.createWithTimeout = createWithTimeout;
})(ShowToastMessageAction || (exports.ShowToastMessageAction = ShowToastMessageAction = {}));
