// @eclipse-glsp/client/lib/features/tools/deletion/delete-tool.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var DelKeyDeleteTool_1, MouseDeleteTool_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteToolMouseListener = exports.MouseDeleteTool = exports.DeleteKeyListener = exports.DelKeyDeleteTool = void 0;
/********************************************************************************
 * Copyright (c) 2019-2024 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const sprotty_1 = require("@eclipse-glsp/sprotty");
const inversify_1 = require("inversify");
const css_feedback_1 = require("../../../base/feedback/css-feedback");
const tool_1 = require("../../../base/tool-manager/tool");
const base_tools_1 = require("../base-tools");
/**
 * Deletes selected elements when hitting the `Del` key.
 */
let DelKeyDeleteTool = DelKeyDeleteTool_1 = class DelKeyDeleteTool {
    constructor() {
        this.isEditTool = true;
        this.deleteKeyListener = new DeleteKeyListener();
    }
    get id() {
        return DelKeyDeleteTool_1.ID;
    }
    enable() {
        this.keytool.register(this.deleteKeyListener);
    }
    disable() {
        this.keytool.deregister(this.deleteKeyListener);
    }
};
exports.DelKeyDeleteTool = DelKeyDeleteTool;
DelKeyDeleteTool.ID = 'glsp.delete-keyboard';
__decorate([
    (0, inversify_1.inject)(sprotty_1.KeyTool),
    __metadata("design:type", sprotty_1.KeyTool)
], DelKeyDeleteTool.prototype, "keytool", void 0);
exports.DelKeyDeleteTool = DelKeyDeleteTool = DelKeyDeleteTool_1 = __decorate([
    (0, inversify_1.injectable)()
], DelKeyDeleteTool);
let DeleteKeyListener = class DeleteKeyListener extends sprotty_1.KeyListener {
    keyDown(element, event) {
        if ((0, sprotty_1.matchesKeystroke)(event, 'Delete')) {
            const deleteElementIds = Array.from(element.root.index
                .all()
                .filter(e => (0, sprotty_1.isDeletable)(e) && (0, sprotty_1.isSelectable)(e) && e.selected)
                .filter(e => e.id !== e.root.id)
                .map(e => e.id));
            if (deleteElementIds.length > 0) {
                return [sprotty_1.DeleteElementOperation.create(deleteElementIds)];
            }
        }
        return [];
    }
};
exports.DeleteKeyListener = DeleteKeyListener;
exports.DeleteKeyListener = DeleteKeyListener = __decorate([
    (0, inversify_1.injectable)()
], DeleteKeyListener);
/**
 * Deletes selected elements when clicking on them.
 */
let MouseDeleteTool = MouseDeleteTool_1 = class MouseDeleteTool extends base_tools_1.BaseEditTool {
    constructor() {
        super(...arguments);
        this.deleteToolMouseListener = new DeleteToolMouseListener();
    }
    get id() {
        return MouseDeleteTool_1.ID;
    }
    enable() {
        const cursorFeedback = this.createFeedbackEmitter()
            .add((0, css_feedback_1.cursorFeedbackAction)(css_feedback_1.CursorCSS.ELEMENT_DELETION), (0, css_feedback_1.cursorFeedbackAction)())
            .submit();
        this.toDisposeOnDisable.push(this.mouseTool.registerListener(this.deleteToolMouseListener), cursorFeedback);
    }
};
exports.MouseDeleteTool = MouseDeleteTool;
MouseDeleteTool.ID = 'glsp.delete-mouse';
exports.MouseDeleteTool = MouseDeleteTool = MouseDeleteTool_1 = __decorate([
    (0, inversify_1.injectable)()
], MouseDeleteTool);
let DeleteToolMouseListener = class DeleteToolMouseListener extends sprotty_1.MouseListener {
    mouseUp(target, event) {
        const deletableParent = (0, sprotty_1.findParentByFeature)(target, sprotty_1.isDeletable);
        if (deletableParent === undefined) {
            return [];
        }
        const result = [];
        result.push(sprotty_1.DeleteElementOperation.create([deletableParent.id]));
        if (!(0, sprotty_1.isCtrlOrCmd)(event)) {
            result.push(tool_1.EnableDefaultToolsAction.create());
        }
        return result;
    }
};
exports.DeleteToolMouseListener = DeleteToolMouseListener;
exports.DeleteToolMouseListener = DeleteToolMouseListener = __decorate([
    (0, inversify_1.injectable)()
], DeleteToolMouseListener);
