// @eclipse-glsp/client/lib/base/shortcuts/available-shortcuts-tool.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023-2025 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var AvailableShortcutsTool_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessibleShortcutKeyListener = exports.AvailableShortcutsTool = void 0;
const sprotty_1 = require("@eclipse-glsp/sprotty");
const inversify_1 = require("inversify");
const base_tools_1 = require("../../features/tools/base-tools");
const available_shortcuts_extension_1 = require("./available-shortcuts-extension");
let AvailableShortcutsTool = AvailableShortcutsTool_1 = class AvailableShortcutsTool extends base_tools_1.BaseTool {
    constructor() {
        super(...arguments);
        this.shortcutKeyListener = new AccessibleShortcutKeyListener();
    }
    get id() {
        return AvailableShortcutsTool_1.ID;
    }
    enable() {
        this.toDisposeOnDisable.push(this.keyTool.registerListener(this.shortcutKeyListener));
    }
};
exports.AvailableShortcutsTool = AvailableShortcutsTool;
AvailableShortcutsTool.ID = 'available-shortcuts-tool';
exports.AvailableShortcutsTool = AvailableShortcutsTool = AvailableShortcutsTool_1 = __decorate([
    (0, inversify_1.injectable)()
], AvailableShortcutsTool);
class AccessibleShortcutKeyListener extends sprotty_1.KeyListener {
    constructor() {
        super(...arguments);
        this.token = Symbol(AccessibleShortcutKeyListener.name);
    }
    keyDown(element, event) {
        if (this.matchesActivateShortcutHelpKeystroke(event)) {
            return [sprotty_1.SetUIExtensionVisibilityAction.create({ extensionId: available_shortcuts_extension_1.AvailableShortcutsUIExtension.ID, visible: true })];
        }
        return [];
    }
    matchesActivateShortcutHelpKeystroke(event) {
        return (0, sprotty_1.matchesKeystroke)(event, 'KeyH', 'alt');
    }
}
exports.AccessibleShortcutKeyListener = AccessibleShortcutKeyListener;
