// @eclipse-glsp/protocol/lib/action-protocol/model-layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerLayoutAction = exports.LayoutOperation = exports.ComputedBoundsAction = exports.RequestBoundsAction = void 0;
const type_util_1 = require("../utils/type-util");
const base_protocol_1 = require("./base-protocol");
var RequestBoundsAction;
(function (RequestBoundsAction) {
    RequestBoundsAction.KIND = 'requestBounds';
    function is(object) {
        return base_protocol_1.RequestAction.hasKind(object, RequestBoundsAction.KIND) && (0, type_util_1.hasObjectProp)(object, 'newRoot');
    }
    RequestBoundsAction.is = is;
    function create(newRoot, options = {}) {
        return {
            kind: RequestBoundsAction.KIND,
            requestId: '',
            newRoot,
            ...options
        };
    }
    RequestBoundsAction.create = create;
})(RequestBoundsAction || (exports.RequestBoundsAction = RequestBoundsAction = {}));
var ComputedBoundsAction;
(function (ComputedBoundsAction) {
    ComputedBoundsAction.KIND = 'computedBounds';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, ComputedBoundsAction.KIND) && (0, type_util_1.hasArrayProp)(object, 'bounds');
    }
    ComputedBoundsAction.is = is;
    function create(bounds, options = {}) {
        return {
            kind: ComputedBoundsAction.KIND,
            responseId: '',
            bounds,
            ...options
        };
    }
    ComputedBoundsAction.create = create;
})(ComputedBoundsAction || (exports.ComputedBoundsAction = ComputedBoundsAction = {}));
var LayoutOperation;
(function (LayoutOperation) {
    LayoutOperation.KIND = 'layout';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, LayoutOperation.KIND);
    }
    LayoutOperation.is = is;
    function create(elementIds, options = {}) {
        return {
            kind: LayoutOperation.KIND,
            isOperation: true,
            elementIds,
            ...options
        };
    }
    LayoutOperation.create = create;
})(LayoutOperation || (exports.LayoutOperation = LayoutOperation = {}));
var TriggerLayoutAction;
(function (TriggerLayoutAction) {
    TriggerLayoutAction.KIND = 'triggerLayout';
    function is(action) {
        return base_protocol_1.Action.hasKind(action, TriggerLayoutAction.KIND);
    }
    TriggerLayoutAction.is = is;
    function create(options = {}) {
        return {
            kind: TriggerLayoutAction.KIND,
            ...options
        };
    }
    TriggerLayoutAction.create = create;
})(TriggerLayoutAction || (exports.TriggerLayoutAction = TriggerLayoutAction = {}));
