// @eclipse-glsp/protocol/lib/action-protocol/model-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceModelChangedAction = exports.UpdateModelAction = exports.SetModelAction = exports.RequestModelAction = void 0;
const type_util_1 = require("../utils/type-util");
const base_protocol_1 = require("./base-protocol");
var RequestModelAction;
(function (RequestModelAction) {
    RequestModelAction.KIND = 'requestModel';
    function is(object) {
        return base_protocol_1.RequestAction.hasKind(object, RequestModelAction.KIND);
    }
    RequestModelAction.is = is;
    function create(options = {}) {
        return {
            kind: RequestModelAction.KIND,
            requestId: '',
            ...options
        };
    }
    RequestModelAction.create = create;
})(RequestModelAction || (exports.RequestModelAction = RequestModelAction = {}));
var SetModelAction;
(function (SetModelAction) {
    SetModelAction.KIND = 'setModel';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, SetModelAction.KIND) && (0, type_util_1.hasObjectProp)(object, 'newRoot');
    }
    SetModelAction.is = is;
    function create(newRoot, options = {}) {
        return {
            kind: SetModelAction.KIND,
            responseId: '',
            newRoot,
            ...options
        };
    }
    SetModelAction.create = create;
})(SetModelAction || (exports.SetModelAction = SetModelAction = {}));
var UpdateModelAction;
(function (UpdateModelAction) {
    UpdateModelAction.KIND = 'updateModel';
    function is(action) {
        return base_protocol_1.Action.hasKind(action, UpdateModelAction.KIND) && (0, type_util_1.hasObjectProp)(action, 'newRoot');
    }
    UpdateModelAction.is = is;
    function create(newRoot, options = {}) {
        return {
            kind: UpdateModelAction.KIND,
            newRoot,
            animate: true,
            ...options
        };
    }
    UpdateModelAction.create = create;
})(UpdateModelAction || (exports.UpdateModelAction = UpdateModelAction = {}));
var SourceModelChangedAction;
(function (SourceModelChangedAction) {
    SourceModelChangedAction.KIND = 'sourceModelChanged';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, SourceModelChangedAction.KIND) && (0, type_util_1.hasStringProp)(object, 'sourceModelName');
    }
    SourceModelChangedAction.is = is;
    function create(sourceModelName) {
        return {
            kind: SourceModelChangedAction.KIND,
            sourceModelName: sourceModelName
        };
    }
    SourceModelChangedAction.create = create;
})(SourceModelChangedAction || (exports.SourceModelChangedAction = SourceModelChangedAction = {}));
