// @eclipse-glsp/client/lib/base/tool-manager/tool.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tool = exports.EnableDefaultToolsAction = exports.EnableToolsAction = void 0;
const sprotty_1 = require("@eclipse-glsp/sprotty");
var EnableToolsAction;
(function (EnableToolsAction) {
    EnableToolsAction.KIND = 'enable-tools';
    function is(object) {
        return sprotty_1.Action.hasKind(object, EnableToolsAction.KIND) && (0, sprotty_1.hasArrayProp)(object, 'toolIds');
    }
    EnableToolsAction.is = is;
    function create(toolIds) {
        return {
            kind: EnableToolsAction.KIND,
            toolIds
        };
    }
    EnableToolsAction.create = create;
})(EnableToolsAction || (exports.EnableToolsAction = EnableToolsAction = {}));
var EnableDefaultToolsAction;
(function (EnableDefaultToolsAction) {
    EnableDefaultToolsAction.KIND = 'enable-default-tools';
    function is(object) {
        return sprotty_1.Action.hasKind(object, EnableDefaultToolsAction.KIND);
    }
    EnableDefaultToolsAction.is = is;
    function create() {
        return {
            kind: EnableDefaultToolsAction.KIND
        };
    }
    EnableDefaultToolsAction.create = create;
})(EnableDefaultToolsAction || (exports.EnableDefaultToolsAction = EnableDefaultToolsAction = {}));
var Tool;
(function (Tool) {
    function is(object) {
        return (sprotty_1.AnyObject.is(object) &&
            (0, sprotty_1.hasStringProp)(object, 'id') &&
            (0, sprotty_1.hasFunctionProp)(object, 'enable') &&
            (0, sprotty_1.hasFunctionProp)(object, 'disable') &&
            (0, sprotty_1.hasBooleanProp)(object, 'isEditTool', true));
    }
    Tool.is = is;
})(Tool || (exports.Tool = Tool = {}));
