// @eclipse-glsp/client/lib/features/tools/marquee-selection/marquee-mouse-tool.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var MarqueeMouseTool_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShiftKeyListener = exports.MarqueeMouseListener = exports.MarqueeMouseTool = void 0;
/********************************************************************************
 * Copyright (c) 2021-2024 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const sprotty_1 = require("@eclipse-glsp/sprotty");
const inversify_1 = require("inversify");
const drag_aware_mouse_listener_1 = require("../../../base/drag-aware-mouse-listener");
const css_feedback_1 = require("../../../base/feedback/css-feedback");
const tool_1 = require("../../../base/tool-manager/tool");
const gmodel_util_1 = require("../../../utils/gmodel-util");
const viewpoint_util_1 = require("../../../utils/viewpoint-util");
const base_tools_1 = require("../base-tools");
const marquee_behavior_1 = require("./marquee-behavior");
const marquee_tool_feedback_1 = require("./marquee-tool-feedback");
let MarqueeMouseTool = MarqueeMouseTool_1 = class MarqueeMouseTool extends base_tools_1.BaseEditTool {
    constructor() {
        super(...arguments);
        this.shiftKeyListener = new ShiftKeyListener();
    }
    get id() {
        return MarqueeMouseTool_1.ID;
    }
    enable() {
        this.toDisposeOnDisable.push(this.mouseTool.registerListener(new MarqueeMouseListener(this.editorContext.modelRoot, this.marqueeUtil)), this.keyTool.registerListener(this.shiftKeyListener), this.createFeedbackEmitter().add((0, css_feedback_1.cursorFeedbackAction)(css_feedback_1.CursorCSS.MARQUEE), (0, css_feedback_1.cursorFeedbackAction)()).submit());
    }
    get isEditTool() {
        return false;
    }
};
exports.MarqueeMouseTool = MarqueeMouseTool;
MarqueeMouseTool.ID = 'glsp.marquee-mouse-tool';
__decorate([
    (0, inversify_1.inject)(marquee_behavior_1.MarqueeUtil),
    __metadata("design:type", marquee_behavior_1.MarqueeUtil)
], MarqueeMouseTool.prototype, "marqueeUtil", void 0);
exports.MarqueeMouseTool = MarqueeMouseTool = MarqueeMouseTool_1 = __decorate([
    (0, inversify_1.injectable)()
], MarqueeMouseTool);
class MarqueeMouseListener extends drag_aware_mouse_listener_1.DragAwareMouseListener {
    constructor(root, marqueeUtil) {
        super();
        this.marqueeUtil = marqueeUtil;
        this.isActive = false;
        // pre-calculate all markable node and edges to improve performance
        this.nodes = this.marqueeUtil.getMarkableNodes(root);
        this.edges = this.marqueeUtil.getMarkableEdges(root);
    }
    mouseDown(target, event) {
        this.isActive = true;
        this.marqueeUtil.updateStartPoint((0, viewpoint_util_1.getAbsolutePosition)(target, event));
        if (event.ctrlKey) {
            this.previouslySelected = (0, gmodel_util_1.getMatchingElements)(target.index, (0, sprotty_1.typeGuard)(gmodel_util_1.isSelectableAndBoundsAware, sprotty_1.isSelected)).map(e => e.id);
        }
        return [];
    }
    mouseMove(target, event) {
        this.marqueeUtil.updateCurrentPoint((0, viewpoint_util_1.getAbsolutePosition)(target, event));
        if (this.isActive) {
            const nodeIdsSelected = this.nodes.filter(e => this.marqueeUtil.isMarked(e)).map(e => e.id);
            const edgeIdsSelected = this.edges.filter(e => this.marqueeUtil.isMarked(e)).map(e => e.id);
            const currentSelected = nodeIdsSelected.concat(edgeIdsSelected);
            const selection = currentSelected.concat(this.previouslySelected);
            return [sprotty_1.SelectAction.setSelection(selection), this.marqueeUtil.drawMarqueeAction()];
        }
        return [];
    }
    mouseUp(target, event) {
        this.isActive = false;
        return this.marqueeUtil.isContinuousMode(target, event)
            ? [marquee_tool_feedback_1.RemoveMarqueeAction.create()]
            : [marquee_tool_feedback_1.RemoveMarqueeAction.create(), tool_1.EnableDefaultToolsAction.create()];
    }
}
exports.MarqueeMouseListener = MarqueeMouseListener;
let ShiftKeyListener = class ShiftKeyListener extends sprotty_1.KeyListener {
    keyUp(element, event) {
        if (event.shiftKey) {
            return [];
        }
        return [marquee_tool_feedback_1.RemoveMarqueeAction.create(), tool_1.EnableDefaultToolsAction.create()];
    }
};
exports.ShiftKeyListener = ShiftKeyListener;
exports.ShiftKeyListener = ShiftKeyListener = __decorate([
    (0, inversify_1.injectable)()
], ShiftKeyListener);
