// @eclipse-glsp/protocol/lib/re-exports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSubType = exports.getBasicType = exports.findElement = exports.cloneModel = exports.applyBounds = exports.isZoomable = exports.isScrollable = void 0;
/********************************************************************************
 * Copyright (c) 2023-2024 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var model_1 = require("sprotty-protocol/lib/model");
Object.defineProperty(exports, "isScrollable", { enumerable: true, get: function () { return model_1.isScrollable; } });
Object.defineProperty(exports, "isZoomable", { enumerable: true, get: function () { return model_1.isZoomable; } });
__exportStar(require("sprotty-protocol/lib/utils/async"), exports);
__exportStar(require("sprotty-protocol/lib/utils/geometry"), exports);
__exportStar(require("sprotty-protocol/lib/utils/json"), exports);
var model_utils_1 = require("sprotty-protocol/lib/utils/model-utils");
Object.defineProperty(exports, "applyBounds", { enumerable: true, get: function () { return model_utils_1.applyBounds; } });
Object.defineProperty(exports, "cloneModel", { enumerable: true, get: function () { return model_utils_1.cloneModel; } });
Object.defineProperty(exports, "findElement", { enumerable: true, get: function () { return model_utils_1.findElement; } });
Object.defineProperty(exports, "getBasicType", { enumerable: true, get: function () { return model_utils_1.getBasicType; } });
Object.defineProperty(exports, "getSubType", { enumerable: true, get: function () { return model_utils_1.getSubType; } });
