// @eclipse-glsp/client/lib/features/accessibility/keyboard-grid/action.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyboardGridKeyboardEventAction = exports.KeyboardGridCellSelectedAction = exports.EnableKeyboardGridAction = void 0;
const sprotty_1 = require("@eclipse-glsp/sprotty");
var EnableKeyboardGridAction;
(function (EnableKeyboardGridAction) {
    EnableKeyboardGridAction.KIND = 'enableKeyboardGrid';
    function is(object) {
        return sprotty_1.Action.hasKind(object, EnableKeyboardGridAction.KIND) && (0, sprotty_1.hasObjectProp)(object, 'options');
    }
    EnableKeyboardGridAction.is = is;
    function create(options) {
        return {
            kind: EnableKeyboardGridAction.KIND,
            options
        };
    }
    EnableKeyboardGridAction.create = create;
})(EnableKeyboardGridAction || (exports.EnableKeyboardGridAction = EnableKeyboardGridAction = {}));
var KeyboardGridCellSelectedAction;
(function (KeyboardGridCellSelectedAction) {
    KeyboardGridCellSelectedAction.KIND = 'keyboardGridCellSelectedAction';
    function is(object) {
        return sprotty_1.Action.hasKind(object, KeyboardGridCellSelectedAction.KIND) && (0, sprotty_1.hasObjectProp)(object, 'options');
    }
    KeyboardGridCellSelectedAction.is = is;
    function create(options) {
        return {
            kind: KeyboardGridCellSelectedAction.KIND,
            options
        };
    }
    KeyboardGridCellSelectedAction.create = create;
})(KeyboardGridCellSelectedAction || (exports.KeyboardGridCellSelectedAction = KeyboardGridCellSelectedAction = {}));
var KeyboardGridKeyboardEventAction;
(function (KeyboardGridKeyboardEventAction) {
    KeyboardGridKeyboardEventAction.KIND = 'keyboardGridKeyboardEvent';
    function is(object) {
        return sprotty_1.Action.hasKind(object, KeyboardGridKeyboardEventAction.KIND) && (0, sprotty_1.hasObjectProp)(object, 'options');
    }
    KeyboardGridKeyboardEventAction.is = is;
    function create(options) {
        return {
            kind: KeyboardGridKeyboardEventAction.KIND,
            options
        };
    }
    KeyboardGridKeyboardEventAction.create = create;
})(KeyboardGridKeyboardEventAction || (exports.KeyboardGridKeyboardEventAction = KeyboardGridKeyboardEventAction = {}));
