// @eclipse-glsp/sprotty/lib/api-override.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MouseListener = exports.KeyListener = void 0;
const inversify_1 = require("inversify");
const sprotty_1 = require("sprotty");
let KeyListener = class KeyListener extends sprotty_1.KeyListener {
    keyDown(element, event) {
        return [];
    }
    keyUp(element, event) {
        return [];
    }
};
exports.KeyListener = KeyListener;
exports.KeyListener = KeyListener = __decorate([
    (0, inversify_1.injectable)()
], KeyListener);
let MouseListener = class MouseListener extends sprotty_1.MouseListener {
    mouseOver(target, event) {
        return [];
    }
    mouseOut(target, event) {
        return [];
    }
    mouseEnter(target, event) {
        return [];
    }
    mouseLeave(target, event) {
        return [];
    }
    mouseDown(target, event) {
        return [];
    }
    mouseMove(target, event) {
        return [];
    }
    mouseUp(target, event) {
        return [];
    }
    wheel(target, event) {
        return [];
    }
    doubleClick(target, event) {
        return [];
    }
    contextMenu(target, event) {
        return [];
    }
    dragOver(target, event) {
        return [];
    }
    drop(target, event) {
        return [];
    }
    decorate(vnode, element) {
        return vnode;
    }
};
exports.MouseListener = MouseListener;
exports.MouseListener = MouseListener = __decorate([
    (0, inversify_1.injectable)()
], MouseListener);
