// @eclipse-glsp/sprotty/lib/feature-modules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sprottyDefaultModule = exports.edgeJunctionModule = exports.updateModule = exports.openModule = exports.moveModule = exports.modelSourceModule = exports.fadeModule = exports.expandModule = exports.edgeLayoutModule = exports.edgeIntersectionModule = exports.edgeEditModule = exports.buttonModule = void 0;
/********************************************************************************
 * Copyright (c) 2023-2025 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable no-restricted-imports */
const di_1 = require("@eclipse-glsp/protocol/lib/di");
const di_config_1 = require("sprotty/lib/base/di.config");
exports.sprottyDefaultModule = di_config_1.default;
const di_config_2 = require("sprotty/lib/features/button/di.config");
const di_config_3 = require("sprotty/lib/features/edge-intersection/di.config");
const di_config_4 = require("sprotty/lib/features/edge-junction/di.config");
const di_config_5 = require("sprotty/lib/features/edge-layout/di.config");
const di_config_6 = require("sprotty/lib/features/edit/di.config");
const di_config_7 = require("sprotty/lib/features/expand/di.config");
const di_config_8 = require("sprotty/lib/features/fade/di.config");
const di_config_9 = require("sprotty/lib/features/move/di.config");
const di_config_10 = require("sprotty/lib/features/open/di.config");
const di_config_11 = require("sprotty/lib/features/update/di.config");
const di_config_12 = require("sprotty/lib/model-source/di.config");
exports.buttonModule = new di_1.FeatureModule(di_config_2.default.registry, { featureId: Symbol('button') });
exports.edgeEditModule = new di_1.FeatureModule(di_config_6.edgeEditModule.registry, { featureId: Symbol('edgeEdit') });
exports.edgeIntersectionModule = new di_1.FeatureModule(di_config_3.default.registry, { featureId: Symbol('edgeIntersection') });
exports.edgeLayoutModule = new di_1.FeatureModule(di_config_5.default.registry, { featureId: Symbol('edgeLayout') });
exports.expandModule = new di_1.FeatureModule(di_config_7.default.registry, { featureId: Symbol('expand') });
exports.fadeModule = new di_1.FeatureModule(di_config_8.default.registry, { featureId: Symbol('fade') });
exports.modelSourceModule = new di_1.FeatureModule(di_config_12.default.registry, { featureId: Symbol('modelSource') });
exports.moveModule = new di_1.FeatureModule(di_config_9.default.registry, { featureId: Symbol('move') });
exports.openModule = new di_1.FeatureModule(di_config_10.default.registry, { featureId: Symbol('open') });
exports.updateModule = new di_1.FeatureModule(di_config_11.default.registry, { featureId: Symbol('update') });
exports.edgeJunctionModule = new di_1.FeatureModule(di_config_4.default.registry, { featureId: Symbol('edgeJunction') });
