// @eclipse-glsp/sprotty/lib/re-exports.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023-2025 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportSvgCommand = exports.isExpandable = exports.expandFeature = exports.isEdgeLayoutable = exports.edgeLayoutFeature = exports.checkEdgePlacement = exports.DEFAULT_EDGE_PLACEMENT = exports.DeleteContextMenuItemProvider = exports.ContextMenuProviderRegistry = exports.RevealNamedElementActionProvider = exports.CommandPaletteActionProviderRegistry = exports.GButton = exports.configureButtonHandler = exports.ButtonHandlerRegistry = exports.layoutableChildFeature = exports.layoutContainerFeature = exports.isSizeable = exports.isLayoutableChild = exports.isLayoutContainer = exports.isBoundsAware = exports.isAlignable = exports.getAbsoluteClientBounds = exports.getAbsoluteBounds = exports.findChildrenAtPosition = exports.boundsFeature = exports.alignFeature = exports.GShapeElement = exports.StatefulLayouter = exports.Layouter = exports.FocusFixPostprocessor = exports.PopupMouseTool = exports.MouseTool = exports.MousePositionTracker = exports.KeyTool = exports.createFeatureSet = exports.SModelRegistry = exports.GModelFactory = exports.EMPTY_ROOT = exports.isParent = exports.createRandomId = exports.ModelIndexImpl = exports.GParentElement = exports.GModelRoot = exports.GModelElement = exports.GChildElement = exports.CommandStack = exports.onAction = exports.configureActionHandler = exports.ActionHandlerRegistry = exports.ActionDispatcher = void 0;
exports.GBezierCreateHandleView = exports.GBezierControlHandleView = exports.BezierCurveEdgeView = exports.SGraphImpl = exports.SEdgeImpl = exports.GPort = exports.GNode = exports.GLabel = exports.GGraphIndex = exports.GCompartment = exports.SprottyBringToFrontCommand = exports.GViewportRootElement = exports.SprottySelectCommand = exports.SprottySelectAllCommand = exports.SelectMouseListener = exports.SelectKeyboardListener = exports.GetSelectionCommand = exports.selectFeature = exports.isSelected = exports.isSelectable = exports.isConnectable = exports.getRouteBounds = exports.getAbsoluteRouteBounds = exports.edgeInProgressTargetHandleID = exports.edgeInProgressID = exports.connectableFeature = exports.GRoutingHandle = exports.GRoutableElement = exports.GDanglingAnchor = exports.GConnectableElement = exports.isProjectable = exports.getProjections = exports.getProjectedBounds = exports.getModelBounds = exports.moveFeature = exports.isMoveable = exports.isLocateable = exports.isDecoration = exports.decorationFeature = exports.SIssueMarkerImpl = exports.GDecoration = exports.popupFeature = exports.isHoverable = exports.hoverFeedbackFeature = exports.hasPopupFeature = exports.isFadeable = exports.fadeFeature = exports.SvgExporter = exports.ExportSvgPostprocessor = exports.ExportSvgKeyListener = void 0;
exports.SvgViewportView = exports.RectangularNodeView = exports.EmptyGroupView = exports.RectangularPort = exports.RectangularNode = exports.GShapedPreRenderedElement = exports.GPreRenderedElement = exports.GHtmlRoot = exports.GForeignObjectElement = exports.DiamondNode = exports.CircularPort = exports.CircularNode = exports.SRoutingHandleView = exports.SGraphView = exports.PolylineEdgeViewWithGapsOnIntersections = exports.PolylineEdgeView = exports.JumpingPolylineEdgeView = exports.GLabelView = exports.GCompartmentView = void 0;
/**
 * @eclipse-glsp/protocol
 */
__exportStar(require("@eclipse-glsp/protocol"), exports);
__exportStar(require("@eclipse-glsp/protocol/lib/di"), exports);
/*
 * sprotty
 */
// ------------------ Base ------------------
// Exclude  definition for labeled actions
// export * from 'sprotty/lib/base/actions/action';
// Exclude IActionDispatcher and IActionDispatcherProvider. Exported via api-override module instead
var action_dispatcher_1 = require("sprotty/lib/base/actions/action-dispatcher");
Object.defineProperty(exports, "ActionDispatcher", { enumerable: true, get: function () { return action_dispatcher_1.ActionDispatcher; } });
var action_handler_1 = require("sprotty/lib/base/actions/action-handler");
Object.defineProperty(exports, "ActionHandlerRegistry", { enumerable: true, get: function () { return action_handler_1.ActionHandlerRegistry; } });
Object.defineProperty(exports, "configureActionHandler", { enumerable: true, get: function () { return action_handler_1.configureActionHandler; } });
Object.defineProperty(exports, "onAction", { enumerable: true, get: function () { return action_handler_1.onAction; } });
__exportStar(require("sprotty/lib/base/actions/diagram-locker"), exports);
__exportStar(require("sprotty/lib/base/animations/animation"), exports);
__exportStar(require("sprotty/lib/base/animations/animation-frame-syncer"), exports);
__exportStar(require("sprotty/lib/base/animations/easing"), exports);
__exportStar(require("sprotty/lib/base/commands/command"), exports);
__exportStar(require("sprotty/lib/base/commands/command-registration"), exports);
// Exclude ICommandStack and ICommandStackProvider. Exported via api-override module instead
var command_stack_1 = require("sprotty/lib/base/commands/command-stack");
Object.defineProperty(exports, "CommandStack", { enumerable: true, get: function () { return command_stack_1.CommandStack; } });
__exportStar(require("sprotty/lib/base/commands/command-stack-options"), exports);
__exportStar(require("sprotty/lib/base/features/initialize-canvas"), exports);
__exportStar(require("sprotty/lib/base/features/set-model"), exports);
// Exclude SModelElementImpl as it as exported with augmentation module
var smodel_1 = require("sprotty/lib/base/model/smodel");
Object.defineProperty(exports, "GChildElement", { enumerable: true, get: function () { return smodel_1.SChildElementImpl; } });
Object.defineProperty(exports, "GModelElement", { enumerable: true, get: function () { return smodel_1.SModelElementImpl; } });
Object.defineProperty(exports, "GModelRoot", { enumerable: true, get: function () { return smodel_1.SModelRootImpl; } });
Object.defineProperty(exports, "GParentElement", { enumerable: true, get: function () { return smodel_1.SParentElementImpl; } });
Object.defineProperty(exports, "ModelIndexImpl", { enumerable: true, get: function () { return smodel_1.ModelIndexImpl; } });
Object.defineProperty(exports, "createRandomId", { enumerable: true, get: function () { return smodel_1.createRandomId; } });
Object.defineProperty(exports, "isParent", { enumerable: true, get: function () { return smodel_1.isParent; } });
var smodel_factory_1 = require("sprotty/lib/base/model/smodel-factory");
Object.defineProperty(exports, "EMPTY_ROOT", { enumerable: true, get: function () { return smodel_factory_1.EMPTY_ROOT; } });
Object.defineProperty(exports, "GModelFactory", { enumerable: true, get: function () { return smodel_factory_1.SModelFactory; } });
// exported without alias we extend it in glsp-client to `GModelRegistry`
Object.defineProperty(exports, "SModelRegistry", { enumerable: true, get: function () { return smodel_factory_1.SModelRegistry; } });
Object.defineProperty(exports, "createFeatureSet", { enumerable: true, get: function () { return smodel_factory_1.createFeatureSet; } });
__exportStar(require("sprotty/lib/base/model/smodel-utils"), exports);
__exportStar(require("sprotty/lib/base/ui-extensions/ui-extension"), exports);
__exportStar(require("sprotty/lib/base/ui-extensions/ui-extension-registry"), exports);
__exportStar(require("sprotty/lib/base/views/dom-helper"), exports);
var key_tool_1 = require("sprotty/lib/base/views/key-tool");
Object.defineProperty(exports, "KeyTool", { enumerable: true, get: function () { return key_tool_1.KeyTool; } });
var mouse_tool_1 = require("sprotty/lib/base/views/mouse-tool");
Object.defineProperty(exports, "MousePositionTracker", { enumerable: true, get: function () { return mouse_tool_1.MousePositionTracker; } });
Object.defineProperty(exports, "MouseTool", { enumerable: true, get: function () { return mouse_tool_1.MouseTool; } });
Object.defineProperty(exports, "PopupMouseTool", { enumerable: true, get: function () { return mouse_tool_1.PopupMouseTool; } });
__exportStar(require("sprotty/lib/base/views/thunk-view"), exports);
__exportStar(require("sprotty/lib/base/views/view"), exports);
__exportStar(require("sprotty/lib/base/views/viewer"), exports);
__exportStar(require("sprotty/lib/base/views/viewer-cache"), exports);
__exportStar(require("sprotty/lib/base/views/viewer-options"), exports);
// Exclude IVnodePostprocessor. Exported via api-override module instead
var vnode_postprocessor_1 = require("sprotty/lib/base/views/vnode-postprocessor");
Object.defineProperty(exports, "FocusFixPostprocessor", { enumerable: true, get: function () { return vnode_postprocessor_1.FocusFixPostprocessor; } });
__exportStar(require("sprotty/lib/base/views/vnode-utils"), exports);
// Exclude sprotty types and export augmented GLSP types instead
// export * from 'sprotty/lib/base/types';
// ------------------ Features ------------------
__exportStar(require("sprotty/lib/features/bounds/bounds-manipulation"), exports);
__exportStar(require("sprotty/lib/features/bounds/hidden-bounds-updater"), exports);
var layout_1 = require("sprotty/lib/features/bounds/layout");
Object.defineProperty(exports, "Layouter", { enumerable: true, get: function () { return layout_1.Layouter; } });
Object.defineProperty(exports, "StatefulLayouter", { enumerable: true, get: function () { return layout_1.StatefulLayouter; } });
var model_1 = require("sprotty/lib/features/bounds/model");
Object.defineProperty(exports, "GShapeElement", { enumerable: true, get: function () { return model_1.SShapeElementImpl; } });
Object.defineProperty(exports, "alignFeature", { enumerable: true, get: function () { return model_1.alignFeature; } });
Object.defineProperty(exports, "boundsFeature", { enumerable: true, get: function () { return model_1.boundsFeature; } });
Object.defineProperty(exports, "findChildrenAtPosition", { enumerable: true, get: function () { return model_1.findChildrenAtPosition; } });
Object.defineProperty(exports, "getAbsoluteBounds", { enumerable: true, get: function () { return model_1.getAbsoluteBounds; } });
Object.defineProperty(exports, "getAbsoluteClientBounds", { enumerable: true, get: function () { return model_1.getAbsoluteClientBounds; } });
Object.defineProperty(exports, "isAlignable", { enumerable: true, get: function () { return model_1.isAlignable; } });
Object.defineProperty(exports, "isBoundsAware", { enumerable: true, get: function () { return model_1.isBoundsAware; } });
Object.defineProperty(exports, "isLayoutContainer", { enumerable: true, get: function () { return model_1.isLayoutContainer; } });
Object.defineProperty(exports, "isLayoutableChild", { enumerable: true, get: function () { return model_1.isLayoutableChild; } });
Object.defineProperty(exports, "isSizeable", { enumerable: true, get: function () { return model_1.isSizeable; } });
Object.defineProperty(exports, "layoutContainerFeature", { enumerable: true, get: function () { return model_1.layoutContainerFeature; } });
Object.defineProperty(exports, "layoutableChildFeature", { enumerable: true, get: function () { return model_1.layoutableChildFeature; } });
__exportStar(require("sprotty/lib/features/bounds/views"), exports);
var button_handler_1 = require("sprotty/lib/features/button/button-handler");
Object.defineProperty(exports, "ButtonHandlerRegistry", { enumerable: true, get: function () { return button_handler_1.ButtonHandlerRegistry; } });
Object.defineProperty(exports, "configureButtonHandler", { enumerable: true, get: function () { return button_handler_1.configureButtonHandler; } });
var model_2 = require("sprotty/lib/features/button/model");
Object.defineProperty(exports, "GButton", { enumerable: true, get: function () { return model_2.SButtonImpl; } });
var action_providers_1 = require("sprotty/lib/features/command-palette/action-providers");
Object.defineProperty(exports, "CommandPaletteActionProviderRegistry", { enumerable: true, get: function () { return action_providers_1.CommandPaletteActionProviderRegistry; } });
Object.defineProperty(exports, "RevealNamedElementActionProvider", { enumerable: true, get: function () { return action_providers_1.RevealNamedElementActionProvider; } });
__exportStar(require("sprotty/lib/features/command-palette/command-palette"), exports);
var menu_providers_1 = require("sprotty/lib/features/context-menu/menu-providers");
Object.defineProperty(exports, "ContextMenuProviderRegistry", { enumerable: true, get: function () { return menu_providers_1.ContextMenuProviderRegistry; } });
Object.defineProperty(exports, "DeleteContextMenuItemProvider", { enumerable: true, get: function () { return menu_providers_1.DeleteContextMenuItemProvider; } });
__exportStar(require("sprotty/lib/features/context-menu/mouse-listener"), exports);
__exportStar(require("sprotty/lib/features/edge-layout/di.config"), exports);
__exportStar(require("sprotty/lib/features/edge-layout/edge-layout"), exports);
var model_3 = require("sprotty/lib/features/edge-layout/model");
Object.defineProperty(exports, "DEFAULT_EDGE_PLACEMENT", { enumerable: true, get: function () { return model_3.DEFAULT_EDGE_PLACEMENT; } });
Object.defineProperty(exports, "checkEdgePlacement", { enumerable: true, get: function () { return model_3.checkEdgePlacement; } });
Object.defineProperty(exports, "edgeLayoutFeature", { enumerable: true, get: function () { return model_3.edgeLayoutFeature; } });
Object.defineProperty(exports, "isEdgeLayoutable", { enumerable: true, get: function () { return model_3.isEdgeLayoutable; } });
// Exclude client-side creation features (not supported in GLSP)
// export * from 'sprotty/lib/features/edit/create';
// export * from 'sprotty/lib/features/edit/create-on-drag';
__exportStar(require("sprotty/lib/features/edit/delete"), exports);
__exportStar(require("sprotty/lib/features/edit/edit-label"), exports);
__exportStar(require("sprotty/lib/features/edit/edit-label-ui"), exports);
__exportStar(require("sprotty/lib/features/edit/edit-routing"), exports);
__exportStar(require("sprotty/lib/features/edit/model"), exports);
// export * from 'sprotty/lib/features/edit/reconnect';
__exportStar(require("sprotty/lib/features/expand/expand"), exports);
var model_4 = require("sprotty/lib/features/expand/model");
Object.defineProperty(exports, "expandFeature", { enumerable: true, get: function () { return model_4.expandFeature; } });
Object.defineProperty(exports, "isExpandable", { enumerable: true, get: function () { return model_4.isExpandable; } });
__exportStar(require("sprotty/lib/features/expand/views"), exports);
// Exclude RequestExportSvgAction. Already provided by glsp-protocol
var export_1 = require("sprotty/lib/features/export/export");
Object.defineProperty(exports, "ExportSvgCommand", { enumerable: true, get: function () { return export_1.ExportSvgCommand; } });
Object.defineProperty(exports, "ExportSvgKeyListener", { enumerable: true, get: function () { return export_1.ExportSvgKeyListener; } });
Object.defineProperty(exports, "ExportSvgPostprocessor", { enumerable: true, get: function () { return export_1.ExportSvgPostprocessor; } });
__exportStar(require("sprotty/lib/features/export/model"), exports);
// Exclude ExportSvgAction. Already provided by glsp-protocol
var svg_exporter_1 = require("sprotty/lib/features/export/svg-exporter");
Object.defineProperty(exports, "SvgExporter", { enumerable: true, get: function () { return svg_exporter_1.SvgExporter; } });
__exportStar(require("sprotty/lib/features/fade/fade"), exports);
var model_5 = require("sprotty/lib/features/fade/model");
Object.defineProperty(exports, "fadeFeature", { enumerable: true, get: function () { return model_5.fadeFeature; } });
Object.defineProperty(exports, "isFadeable", { enumerable: true, get: function () { return model_5.isFadeable; } });
__exportStar(require("sprotty/lib/features/hover/hover"), exports);
var model_6 = require("sprotty/lib/features/hover/model");
Object.defineProperty(exports, "hasPopupFeature", { enumerable: true, get: function () { return model_6.hasPopupFeature; } });
Object.defineProperty(exports, "hoverFeedbackFeature", { enumerable: true, get: function () { return model_6.hoverFeedbackFeature; } });
Object.defineProperty(exports, "isHoverable", { enumerable: true, get: function () { return model_6.isHoverable; } });
Object.defineProperty(exports, "popupFeature", { enumerable: true, get: function () { return model_6.popupFeature; } });
__exportStar(require("sprotty/lib/features/hover/popup-position-updater"), exports);
// Alias SModel types
__exportStar(require("sprotty/lib/features/decoration/decoration-placer"), exports);
var model_7 = require("sprotty/lib/features/decoration/model");
Object.defineProperty(exports, "GDecoration", { enumerable: true, get: function () { return model_7.SDecoration; } });
Object.defineProperty(exports, "SIssueMarkerImpl", { enumerable: true, get: function () { return model_7.SIssueMarkerImpl; } });
Object.defineProperty(exports, "decorationFeature", { enumerable: true, get: function () { return model_7.decorationFeature; } });
Object.defineProperty(exports, "isDecoration", { enumerable: true, get: function () { return model_7.isDecoration; } });
__exportStar(require("sprotty/lib/features/decoration/views"), exports);
__exportStar(require("sprotty/lib/features/edge-intersection/intersection-finder"), exports);
__exportStar(require("sprotty/lib/features/edge-intersection/sweepline"), exports);
__exportStar(require("sprotty/lib/features/edge-junction/junction-finder"), exports);
var model_8 = require("sprotty/lib/features/move/model");
Object.defineProperty(exports, "isLocateable", { enumerable: true, get: function () { return model_8.isLocateable; } });
Object.defineProperty(exports, "isMoveable", { enumerable: true, get: function () { return model_8.isMoveable; } });
Object.defineProperty(exports, "moveFeature", { enumerable: true, get: function () { return model_8.moveFeature; } });
__exportStar(require("sprotty/lib/features/move/move"), exports);
__exportStar(require("sprotty/lib/features/move/snap"), exports);
__exportStar(require("sprotty/lib/features/nameable/model"), exports);
__exportStar(require("sprotty/lib/features/open/model"), exports);
__exportStar(require("sprotty/lib/features/open/open"), exports);
var model_9 = require("sprotty/lib/features/projection/model");
Object.defineProperty(exports, "getModelBounds", { enumerable: true, get: function () { return model_9.getModelBounds; } });
Object.defineProperty(exports, "getProjectedBounds", { enumerable: true, get: function () { return model_9.getProjectedBounds; } });
Object.defineProperty(exports, "getProjections", { enumerable: true, get: function () { return model_9.getProjections; } });
Object.defineProperty(exports, "isProjectable", { enumerable: true, get: function () { return model_9.isProjectable; } });
__exportStar(require("sprotty/lib/features/projection/views"), exports);
__exportStar(require("sprotty/lib/features/routing/abstract-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/anchor"), exports);
__exportStar(require("sprotty/lib/features/routing/bezier-anchors"), exports);
__exportStar(require("sprotty/lib/features/routing/bezier-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/manhattan-anchors"), exports);
__exportStar(require("sprotty/lib/features/routing/manhattan-edge-router"), exports);
// Alias SModel types
var model_10 = require("sprotty/lib/features/routing/model");
Object.defineProperty(exports, "GConnectableElement", { enumerable: true, get: function () { return model_10.SConnectableElementImpl; } });
Object.defineProperty(exports, "GDanglingAnchor", { enumerable: true, get: function () { return model_10.SDanglingAnchorImpl; } });
Object.defineProperty(exports, "GRoutableElement", { enumerable: true, get: function () { return model_10.SRoutableElementImpl; } });
Object.defineProperty(exports, "GRoutingHandle", { enumerable: true, get: function () { return model_10.SRoutingHandleImpl; } });
Object.defineProperty(exports, "connectableFeature", { enumerable: true, get: function () { return model_10.connectableFeature; } });
Object.defineProperty(exports, "edgeInProgressID", { enumerable: true, get: function () { return model_10.edgeInProgressID; } });
Object.defineProperty(exports, "edgeInProgressTargetHandleID", { enumerable: true, get: function () { return model_10.edgeInProgressTargetHandleID; } });
Object.defineProperty(exports, "getAbsoluteRouteBounds", { enumerable: true, get: function () { return model_10.getAbsoluteRouteBounds; } });
Object.defineProperty(exports, "getRouteBounds", { enumerable: true, get: function () { return model_10.getRouteBounds; } });
Object.defineProperty(exports, "isConnectable", { enumerable: true, get: function () { return model_10.isConnectable; } });
__exportStar(require("sprotty/lib/features/routing/polyline-anchors"), exports);
__exportStar(require("sprotty/lib/features/routing/polyline-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/routing"), exports);
__exportStar(require("sprotty/lib/features/routing/views"), exports);
var model_11 = require("sprotty/lib/features/select/model");
Object.defineProperty(exports, "isSelectable", { enumerable: true, get: function () { return model_11.isSelectable; } });
Object.defineProperty(exports, "isSelected", { enumerable: true, get: function () { return model_11.isSelected; } });
Object.defineProperty(exports, "selectFeature", { enumerable: true, get: function () { return model_11.selectFeature; } });
// Alias Select commands with sprotty prefix to avoid clash with glsp-client
var select_1 = require("sprotty/lib/features/select/select");
Object.defineProperty(exports, "GetSelectionCommand", { enumerable: true, get: function () { return select_1.GetSelectionCommand; } });
Object.defineProperty(exports, "SelectKeyboardListener", { enumerable: true, get: function () { return select_1.SelectKeyboardListener; } });
Object.defineProperty(exports, "SelectMouseListener", { enumerable: true, get: function () { return select_1.SelectMouseListener; } });
Object.defineProperty(exports, "SprottySelectAllCommand", { enumerable: true, get: function () { return select_1.SelectAllCommand; } });
Object.defineProperty(exports, "SprottySelectCommand", { enumerable: true, get: function () { return select_1.SelectCommand; } });
// Exclude undo-redo as the client provides its own implementation
// export * from 'sprotty/lib/features/undo-redo/undo-redo';
__exportStar(require("sprotty/lib/features/update/model-matching"), exports);
__exportStar(require("sprotty/lib/features/update/update-model"), exports);
__exportStar(require("sprotty/lib/features/viewport/center-fit"), exports);
__exportStar(require("sprotty/lib/features/viewport/model"), exports);
__exportStar(require("sprotty/lib/features/viewport/scroll"), exports);
// Alias SModel types
__exportStar(require("sprotty/lib/features/viewport/viewport"), exports);
var viewport_root_1 = require("sprotty/lib/features/viewport/viewport-root");
Object.defineProperty(exports, "GViewportRootElement", { enumerable: true, get: function () { return viewport_root_1.ViewportRootElementImpl; } });
__exportStar(require("sprotty/lib/features/viewport/zoom"), exports);
var zorder_1 = require("sprotty/lib/features/zorder/zorder");
Object.defineProperty(exports, "SprottyBringToFrontCommand", { enumerable: true, get: function () { return zorder_1.BringToFrontCommand; } });
// ------------------ Graph ------------------
// Alias SModel types
var sgraph_1 = require("sprotty/lib/graph/sgraph");
Object.defineProperty(exports, "GCompartment", { enumerable: true, get: function () { return sgraph_1.SCompartmentImpl; } });
Object.defineProperty(exports, "GGraphIndex", { enumerable: true, get: function () { return sgraph_1.SGraphIndex; } });
Object.defineProperty(exports, "GLabel", { enumerable: true, get: function () { return sgraph_1.SLabelImpl; } });
Object.defineProperty(exports, "GNode", { enumerable: true, get: function () { return sgraph_1.SNodeImpl; } });
Object.defineProperty(exports, "GPort", { enumerable: true, get: function () { return sgraph_1.SPortImpl; } });
// Exported as is, we extend it in glsp-client to GEdge
Object.defineProperty(exports, "SEdgeImpl", { enumerable: true, get: function () { return sgraph_1.SEdgeImpl; } });
// Exported as is, we extend it in glsp-client to GGraph
Object.defineProperty(exports, "SGraphImpl", { enumerable: true, get: function () { return sgraph_1.SGraphImpl; } });
// Alias SModel types
var views_1 = require("sprotty/lib/graph/views");
Object.defineProperty(exports, "BezierCurveEdgeView", { enumerable: true, get: function () { return views_1.BezierCurveEdgeView; } });
Object.defineProperty(exports, "GBezierControlHandleView", { enumerable: true, get: function () { return views_1.SBezierControlHandleView; } });
Object.defineProperty(exports, "GBezierCreateHandleView", { enumerable: true, get: function () { return views_1.SBezierCreateHandleView; } });
Object.defineProperty(exports, "GCompartmentView", { enumerable: true, get: function () { return views_1.SCompartmentView; } });
Object.defineProperty(exports, "GLabelView", { enumerable: true, get: function () { return views_1.SLabelView; } });
Object.defineProperty(exports, "JumpingPolylineEdgeView", { enumerable: true, get: function () { return views_1.JumpingPolylineEdgeView; } });
Object.defineProperty(exports, "PolylineEdgeView", { enumerable: true, get: function () { return views_1.PolylineEdgeView; } });
Object.defineProperty(exports, "PolylineEdgeViewWithGapsOnIntersections", { enumerable: true, get: function () { return views_1.PolylineEdgeViewWithGapsOnIntersections; } });
Object.defineProperty(exports, "SGraphView", { enumerable: true, get: function () { return views_1.SGraphView; } });
Object.defineProperty(exports, "SRoutingHandleView", { enumerable: true, get: function () { return views_1.SRoutingHandleView; } });
// ------------------ Library ------------------
__exportStar(require("sprotty/lib/lib/generic-views"), exports);
__exportStar(require("sprotty/lib/lib/html-views"), exports);
__exportStar(require("sprotty/lib/lib/jsx"), exports);
// Alias SModel types
var model_12 = require("sprotty/lib/lib/model");
Object.defineProperty(exports, "CircularNode", { enumerable: true, get: function () { return model_12.CircularNode; } });
Object.defineProperty(exports, "CircularPort", { enumerable: true, get: function () { return model_12.CircularPort; } });
Object.defineProperty(exports, "DiamondNode", { enumerable: true, get: function () { return model_12.DiamondNode; } });
Object.defineProperty(exports, "GForeignObjectElement", { enumerable: true, get: function () { return model_12.ForeignObjectElementImpl; } });
Object.defineProperty(exports, "GHtmlRoot", { enumerable: true, get: function () { return model_12.HtmlRootImpl; } });
Object.defineProperty(exports, "GPreRenderedElement", { enumerable: true, get: function () { return model_12.PreRenderedElementImpl; } });
Object.defineProperty(exports, "GShapedPreRenderedElement", { enumerable: true, get: function () { return model_12.ShapedPreRenderedElementImpl; } });
Object.defineProperty(exports, "RectangularNode", { enumerable: true, get: function () { return model_12.RectangularNode; } });
Object.defineProperty(exports, "RectangularPort", { enumerable: true, get: function () { return model_12.RectangularPort; } });
__exportStar(require("sprotty/lib/lib/modules"), exports);
var svg_views_1 = require("sprotty/lib/lib/svg-views");
Object.defineProperty(exports, "EmptyGroupView", { enumerable: true, get: function () { return svg_views_1.EmptyGroupView; } });
Object.defineProperty(exports, "RectangularNodeView", { enumerable: true, get: function () { return svg_views_1.RectangularNodeView; } });
Object.defineProperty(exports, "SvgViewportView", { enumerable: true, get: function () { return svg_views_1.SvgViewportView; } });
// ------------------ Model Source ------------------
__exportStar(require("sprotty/lib/model-source/commit-model"), exports);
// Exclude as not supported in GLSP
// export * from 'sprotty/lib/model-source/diagram-server';
// export * from 'sprotty/lib/model-source/local-model-source';
__exportStar(require("sprotty/lib/model-source/logging"), exports);
__exportStar(require("sprotty/lib/model-source/model-source"), exports);
// export * from 'sprotty/lib/model-source/websocket';
// ------------------ Utilities ------------------
__exportStar(require("sprotty/lib/utils/browser"), exports);
__exportStar(require("sprotty/lib/utils/codicon"), exports);
__exportStar(require("sprotty/lib/utils/color"), exports);
__exportStar(require("sprotty/lib/utils/geometry"), exports);
__exportStar(require("sprotty/lib/utils/inversify"), exports);
__exportStar(require("sprotty/lib/utils/iterable"), exports);
__exportStar(require("sprotty/lib/utils/keyboard"), exports);
__exportStar(require("sprotty/lib/utils/logging"), exports);
__exportStar(require("sprotty/lib/utils/registry"), exports);
