// @eclipse-glsp/client/lib/features/label-edit/edit-label-validator.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BalloonLabelValidationDecorator = exports.ServerEditLabelValidator = exports.LabelEditValidation = void 0;
/********************************************************************************
 * Copyright (c) 2019-2024 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const sprotty_1 = require("@eclipse-glsp/sprotty");
const inversify_1 = require("inversify");
var LabelEditValidation;
(function (LabelEditValidation) {
    LabelEditValidation.CONTEXT_ID = 'label-edit';
    function toEditLabelValidationResult(status) {
        const message = status.message;
        let severity = 'ok';
        if (sprotty_1.ValidationStatus.isError(status)) {
            severity = 'error';
        }
        else if (sprotty_1.ValidationStatus.isWarning(status)) {
            severity = 'warning';
        }
        return { message, severity };
    }
    LabelEditValidation.toEditLabelValidationResult = toEditLabelValidationResult;
    function createValidationRequestAction(value, labelId) {
        return sprotty_1.RequestEditValidationAction.create({ contextId: LabelEditValidation.CONTEXT_ID, modelElementId: labelId, text: value });
    }
    LabelEditValidation.createValidationRequestAction = createValidationRequestAction;
})(LabelEditValidation || (exports.LabelEditValidation = LabelEditValidation = {}));
let ServerEditLabelValidator = class ServerEditLabelValidator {
    async validate(value, label) {
        const action = LabelEditValidation.createValidationRequestAction(value, label.id);
        const response = await this.actionDispatcher.requestUntil(action);
        return response ? this.getValidationResultFromResponse(response) : { severity: 'ok' };
    }
    getValidationResultFromResponse(action) {
        if (sprotty_1.SetEditValidationResultAction.is(action)) {
            return LabelEditValidation.toEditLabelValidationResult(action.status);
        }
        return { severity: 'ok' };
    }
};
exports.ServerEditLabelValidator = ServerEditLabelValidator;
__decorate([
    (0, inversify_1.inject)(sprotty_1.TYPES.IActionDispatcher),
    __metadata("design:type", Object)
], ServerEditLabelValidator.prototype, "actionDispatcher", void 0);
exports.ServerEditLabelValidator = ServerEditLabelValidator = __decorate([
    (0, inversify_1.injectable)()
], ServerEditLabelValidator);
let BalloonLabelValidationDecorator = class BalloonLabelValidationDecorator {
    decorate(input, result) {
        const containerElement = input.parentElement;
        if (!containerElement) {
            return;
        }
        if (result.message) {
            containerElement.setAttribute('data-balloon', result.message);
            containerElement.setAttribute('data-balloon-pos', 'up-left');
            containerElement.setAttribute('data-balloon-visible', 'true');
        }
        switch (result.severity) {
            case 'ok':
                containerElement.classList.add('validation-ok');
                break;
            case 'warning':
                containerElement.classList.add('validation-warning');
                break;
            case 'error':
                containerElement.classList.add('validation-error');
                break;
        }
    }
    dispose(input) {
        const containerElement = input.parentElement;
        if (containerElement) {
            containerElement.removeAttribute('data-balloon');
            containerElement.removeAttribute('data-balloon-pos');
            containerElement.removeAttribute('data-balloon-visible');
            containerElement.classList.remove('validation-ok', 'validation-warning', 'validation-error');
        }
    }
};
exports.BalloonLabelValidationDecorator = BalloonLabelValidationDecorator;
exports.BalloonLabelValidationDecorator = BalloonLabelValidationDecorator = __decorate([
    (0, inversify_1.injectable)()
], BalloonLabelValidationDecorator);
