// @eclipse-glsp/client/lib/features/bounds/layout-data.js
"use strict";
/********************************************************************************
 * Copyright (c) 2024 Axon Ivy AG and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.LayoutAware = void 0;
var LayoutAware;
(function (LayoutAware) {
    function is(element) {
        return 'layoutData' in element;
    }
    LayoutAware.is = is;
    function getLayoutData(element) {
        return is(element) ? element.layoutData : undefined;
    }
    LayoutAware.getLayoutData = getLayoutData;
    function setLayoutData(element, data) {
        element.layoutData = data;
    }
    LayoutAware.setLayoutData = setLayoutData;
    function setComputedDimensions(element, computedDimensions) {
        ensureLayoutAware(element).layoutData.computedDimensions = computedDimensions;
    }
    LayoutAware.setComputedDimensions = setComputedDimensions;
    function getComputedDimensions(element) {
        var _a;
        return (_a = getLayoutData(element)) === null || _a === void 0 ? void 0 : _a.computedDimensions;
    }
    LayoutAware.getComputedDimensions = getComputedDimensions;
    function ensureLayoutAware(element) {
        var _a;
        element.layoutData = (_a = element.layoutData) !== null && _a !== void 0 ? _a : {};
        return element;
    }
})(LayoutAware || (exports.LayoutAware = LayoutAware = {}));
