// @eclipse-glsp/client/lib/utils/geometry-util.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023-2024 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAfter = exports.isBefore = exports.isBelow = exports.isAbove = exports.compareFunction = exports.sortBy = exports.bottomRight = exports.bottomCenter = exports.bottomLeft = exports.middleRight = exports.middleCenter = exports.middleLeft = exports.topRight = exports.topCenter = exports.topLeft = exports.bottom = exports.middle = exports.top = exports.right = exports.center = exports.left = void 0;
const sprotty_1 = require("@eclipse-glsp/sprotty");
/** @deprecated Use {@link Bounds.left} */
exports.left = sprotty_1.Bounds.left;
/** @deprecated Use {@link Bounds.centerX} */
exports.center = sprotty_1.Bounds.center;
/** @deprecated Use {@link Bounds.right} */
exports.right = sprotty_1.Bounds.right;
/** @deprecated Use {@link Bounds.top} */
exports.top = sprotty_1.Bounds.top;
/** @deprecated Use {@link Bounds.middle} */
exports.middle = sprotty_1.Bounds.middle;
/** @deprecated Use {@link Bounds.bottom} */
exports.bottom = sprotty_1.Bounds.bottom;
/** @deprecated Use {@link Bounds.topLeft} */
exports.topLeft = sprotty_1.Bounds.topLeft;
/** @deprecated Use {@link Bounds.topCenter} */
exports.topCenter = sprotty_1.Bounds.topCenter;
/** @deprecated Use {@link Bounds.topRight} */
exports.topRight = sprotty_1.Bounds.topRight;
/** @deprecated Use {@link Bounds.middleLeft} */
exports.middleLeft = sprotty_1.Bounds.middleLeft;
/** @deprecated Use {@link Bounds.middleCenter} */
exports.middleCenter = sprotty_1.Bounds.middleCenter;
/** @deprecated Use {@link Bounds.middleRight} */
exports.middleRight = sprotty_1.Bounds.middleRight;
/** @deprecated Use {@link Bounds.bottomLeft} */
exports.bottomLeft = sprotty_1.Bounds.bottomLeft;
/** @deprecated Use {@link Bounds.bottomCenter} */
exports.bottomCenter = sprotty_1.Bounds.bottomCenter;
/** @deprecated Use {@link Bounds.bottomRight} */
exports.bottomRight = sprotty_1.Bounds.bottomRight;
/** @deprecated Use {@link Bounds.left} */
exports.sortBy = sprotty_1.Bounds.sortBy;
const compareFunction = (rankFunc) => (x, y) => rankFunc(x) - rankFunc(y);
exports.compareFunction = compareFunction;
/** @deprecated Use {@link Bounds.isAbove} */
exports.isAbove = sprotty_1.Bounds.isAbove;
/** @deprecated Use {@link Bounds.isBelow} */
exports.isBelow = sprotty_1.Bounds.isBelow;
/** @deprecated Use {@link Bounds.isBefore} */
exports.isBefore = sprotty_1.Bounds.isBefore;
/** @deprecated Use {@link Bounds.isAfter} */
exports.isAfter = sprotty_1.Bounds.isAfter;
